### PR TITLE
MSEARCH-14 Implement CQL query parser

### PR DESCRIPTION
**Pull request by pfilippovEPAM**
Implementation of CQL query parser with required endpoint, service, converter and repository to perform search operation to the Elasticsearch server.
Currently not all features supported that are already done in existing inventory service, however they will be implemented in the following stories.

### CODE REVIEW

File: src/main/java/org/folio/search/cql/CqlSearchQueryConverter.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        return boolQuery
+          .should(convertToQuery(node.getLeftOperand()))
+          .should(convertToQuery(node.getRightOperand()));
+      case AND:
+        return boolQuery
+          .must(convertToQuery(node.getLeftOperand()))
+          .must(convertToQuery(node.getRightOperand()));
+      case NOT:
+        return boolQuery
+          .must(convertToQuery(node.getLeftOperand()))
```
Comment (MikhailFokanov) [inline src/main/java/org/folio/search/cql/CqlSearchQueryConverter.java:132]: I cannot get how does it work... Did you test it with elastic?

**Review by MikhailFokanov — APPROVED** [review]
Looks good.

**Review by bohdan-suprun — APPROVED** [review]

File: src/main/java/org/folio/search/model/rest/response/PagedResult.java [hunk 2]
```diff
@@ -1,18 +1,10 @@
...
 
-  private long total;
-  private long totalPages;
+  long getTotalRecords();
 
-  private int pageSize;
-  private int pageNumber;
-
-  private List<T> content;
+  List<T> getInstances();
```
Comment (hjiebsco) [inline src/main/java/org/folio/search/model/rest/response/PagedResult.java:9]: Assuming the result could be instances, holdings, or items, maybe consider renaming to `getRecords` to be more generic?

File: src/main/java/org/folio/search/model/rest/response/SearchResult.java [hunk 3]
```diff
@@ -1,11 +1,14 @@
...
 
 @Data
-@NoArgsConstructor
-@EqualsAndHashCode(callSuper = true)
-public class SearchResult extends PagedResult<Object> {
+@AllArgsConstructor(staticName = "of")
+public class SearchResult implements PagedResult<Map<String, Object>> {
+
+  private final long totalRecords;
+  private final List<Map<String, Object>> instances;
```
Comment (hjiebsco) [inline src/main/java/org/folio/search/model/rest/response/SearchResult.java:13]: Consider renaming to `records` if the same generic idea applies.

File: src/main/java/org/folio/search/repository/SearchRepository.java [hunk 4]
```diff
@@ -1,24 +1,62 @@
...
+import org.elasticsearch.search.builder.SearchSourceBuilder;
 import org.folio.search.model.rest.response.SearchResult;
+import org.folio.search.model.service.CqlSearchRequest;
 import org.springframework.stereotype.Repository;
 
 /**
  * Search resource repository with set of operation to perform search operations.
  */
 @Repository
+@RequiredArgsConstructor
```
Comment (hjiebsco) [inline src/main/java/org/folio/search/repository/SearchRepository.java:26]: Consider to do the same for `IndexRepository.java`

**Review by hjiebsco — APPROVED** [review]
Looks good. Left a few minor comments. Thanks.

File: src/main/java/org/folio/search/model/rest/response/SearchResult.java (hunk 3 above)
Comment (pfilippovEPAM) [reply src/main/java/org/folio/search/model/rest/response/SearchResult.java:13]: In current implementation - I am trying to support the contract, that is used in existing inventory search and it returns only instances
Also, during working on swagger documentation - map has been replaced with instance object

File: src/main/java/org/folio/search/repository/SearchRepository.java (hunk 4 above)
Comment (pfilippovEPAM) [reply src/main/java/org/folio/search/repository/SearchRepository.java:26]: fixed

File: pom.xml [hunk 5]
```diff
@@ -111,6 +111,22 @@
       <version>2.4.0</version>
     </dependency>
 
+    <dependency>
+      <groupId>org.folio</groupId>
+      <artifactId>cql2pgjson</artifactId>
+      <version>29.5.0-SNAPSHOT</version>
```
[the remark this answers is not in the document]
Comment (pfilippovEPAM) [reply pom.xml]: changed to the latest version

File: pom.xml [hunk 6]
```diff
@@ -264,6 +280,11 @@
       <name>FOLIO Maven repository</name>
       <url>https://repository.folio.org/repository/maven-folio</url>
     </repository>
+    <repository>
+      <id>indexdata</id>
+      <name>Index Data</name>
+      <url>http://maven.indexdata.com/</url>
+    </repository>
```
[the remark this answers is not in the document]
Comment (pfilippovEPAM) [reply pom.xml]: Yes, thank you, it's not used with new version

**Pull request merged**